### ISSUE TITLE
Issues in replicating the model in Tensorflow

### ISSUE THREAD

**Issue by sarthmit** [enalisnick/stick-breaking_dgms#1]
Hi!
I am trying to replicate the results of a standard Stick Breaking VAE (unsupervised) in Tensorflow. One issue I found was even after using the same prior as in your Theano code, the KL divergence came out negative and then the optimizer rapidly reduced it to large to negative values and finally NaN. This corresponded to very small values of alpha and beta (around 0.001s or lower).

I solved that problem by clipping alpha and beta to between 0.25 and 15. However, am not getting good sampled images on using 50 dimensional latent space. I am noticing that though the negative log likelihood i reducing, the KL divergence increases from about 1 to 4. Did you encounter similar issues and if so, what did you do to solve them?

**Comment by enalisnick** [comment]
Hi, thanks for reaching out.  

Yes, there are some numerical instabilities, mostly from the approximations needed for the KLD computation.  Did you see the "Computational Issues to Note" section of the ReadMe?  Most of the difficulties I've encountered are discussed there.  

I've found most problems can be solved by increasing the truncation size.  When the truncation is too low, the model tries to do a kind of hard thresholding (as in trying to strictly divide different digit classes to different dimensionalities).  I would try that first.  

How you initialize the weights that parametrize the Kumar. a and b parameters also can be important.  If they are set too high or too low, then there are some overflow risks in the sampling process.  Clipping should solve most problems but I'd say that a min value of 0.25 is a little too large.    

Lastly, the KLD should increase as training progresses.  This is appropriate behavior since the latent space should be moving away from the prior and moving toward a configuration that can better represent the data.  The ELBO should still decrease, though, since the log likelihood term  should be decreasing more than the KLD is increasing.

**Comment by enalisnick** [comment]
Also, you may want to check out [this repository](https://github.com/enalisnick/mixture_density_VAEs/blob/master/models/gaussMMVAE_collapsed.py) in which I implement a full Dirichlet process latent space in Tensorflow.  It contains the same stick-breaking code ported to TF.

**Comment by sarthmit** [comment]
I am setting the prior parameters to the same value as in your theano code. I am also using truncation size as 50 as mentioned in the paper but am still not getting good samples. For clipping, what sort of min value would you suggest?

I did go through the computational issues to Note section but couldn't solve the problem. Currently the model I am training reaches to around 6 KL Divergence and after training for some epochs (around 400-500), I am getting these types of samples:

![plot](https://user-images.githubusercontent.com/19748754/38830965-8c97ba9c-41db-11e8-90a1-4092fe35e9b0.png)

I will check out that repository asap :)
Any other help/advice is greatly appreciated and thank you for replying!

**Comment by enalisnick** [comment]
Maybe you are just not training it long enough?  I think the samples in the paper are from a model trained for about 1,150 epochs.  Here's the log output for when it reached convergence (adam lr of 3e-4):

> epoch 1150, neg. log likelihood 80.2558137004, kl divergence 21.3541852125, ELBO 101.609999059, avg latent dim. 46.7028231293, test NLL 81.7177581787
